### PR TITLE
 CBG-356: Backward compatibility for cache config

### PR DESCRIPTION
**Pull request by JRascagneres**
This is based off of CBG-355 and therefore that will need to be reviewed and merged first and then this will need to be rebased.
- [x] Merge CBG-355
- [x] Rebase this branch

### CODE REVIEW

File: rest/config.go [hunk 1]
```diff
@@ -461,6 +504,76 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
...
+// Checks for deprecated cache config options and if they are set it will return a warning. If the old one is set and
+// the new one is not set it will set the new to the old value. If they are both set it will still give the warning but
+// will choose the new value.
+func (dbConfig *DbConfig) deprecatedConfigCacheFallback() (warnings []string) {
+
+	warningMsgFmt := "Using deprecated config option: %q. Use %q instead."
+
+	if dbConfig.DeprecatedRevCacheSize != nil {
+		if dbConfig.CacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig.Size == nil {
+			dbConfig.CacheConfig.RevCacheConfig.Size = dbConfig.DeprecatedRevCacheSize
```
Comment (bbrks) [inline rest/config.go:532]: I think we'll get a nil pointer panic here when `dbConfig.CacheConfig` or `dbConfig.CacheConfig.RevCacheConfig` are nil.

We'll probably need to instantiate those structs here when they're not there already.

File: rest/config.go [hunk 2]
```diff
@@ -461,6 +504,76 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
...
+		}
+		warnings = append(warnings, fmt.Sprintf(warningMsgFmt, "rev_cache_size", "cache.rev_cache.size."))
+	}
+
+	if dbConfig.CacheConfig != nil {
+
+		if dbConfig.CacheConfig.DeprecatedCachePendingSeqMaxWait != nil {
+			if dbConfig.CacheConfig.ChannelCacheConfig == nil || dbConfig.CacheConfig.ChannelCacheConfig.MaxWaitPending == nil {
+				dbConfig.CacheConfig.ChannelCacheConfig.MaxWaitPending = dbConfig.CacheConfig.DeprecatedCachePendingSeqMaxWait
```
Comment (bbrks) [inline rest/config.go]: > I think we'll get a nil pointer panic here when `dbConfig.CacheConfig` or `dbConfig.CacheConfig.RevCacheConfig` are nil.
> 
> We'll probably need to instantiate those structs here when they're not there already.

Same with all of these.

File: rest/config.go [hunk 3]
```diff
@@ -461,6 +504,76 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
...
+			}
+			warnings = append(warnings, fmt.Sprintf(warningMsgFmt, "max_wait_pending", "channel_cache.max_wait_pending"))
```
Comment (bbrks) [inline rest/config.go]: Do we need to prefix the new options of these warnings with `cache.`?

File: rest/config.go [hunk 4]
```diff
@@ -461,6 +504,76 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
...
+// will choose the new value.
+func (dbConfig *DbConfig) deprecatedConfigCacheFallback() (warnings []string) {
+
+	warningMsgFmt := "Using deprecated config option: %q. Use %q instead."
+
+	if dbConfig.DeprecatedRevCacheSize != nil {
+		if dbConfig.CacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig.Size == nil {
+			dbConfig.CacheConfig.RevCacheConfig.Size = dbConfig.DeprecatedRevCacheSize
+		}
+		warnings = append(warnings, fmt.Sprintf(warningMsgFmt, "rev_cache_size", "cache.rev_cache.size."))
```
Comment (bbrks) [inline rest/config.go]: Looks like an extra `.` was left in

```suggestion
		warnings = append(warnings, fmt.Sprintf(warningMsgFmt, "rev_cache_size", "cache.rev_cache.size"))
```

**Review by bbrks — CHANGES_REQUESTED** [review]

File: rest/config.go (hunk 3 above)
Comment (JRascagneres) [reply rest/config.go]: I omitted this so it was relative to where the old option was vs the new but makes more sense to have the full path.

File: rest/config.go [hunk 5]
```diff
@@ -461,6 +508,88 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
 
 }
 
+// Checks for deprecated cache config options and if they are set it will return a warning. If the old one is set and
+// the new one is not set it will set the new to the old value. If they are both set it will still give the warning but
+// will choose the new value.
+func (dbConfig *DbConfig) deprecatedConfigCacheFallback() (warnings []string) {
```
Comment (adamcfraser) [inline rest/config.go:514]: Include some unit tests specifically for this function.

File: rest/config.go [hunk 6]
```diff
@@ -461,6 +508,88 @@ func (dbConfig *DbConfig) validateSgDbConfig() error {
...
+	}
+
+	if dbConfig.DeprecatedRevCacheSize != nil {
+		if dbConfig.CacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig == nil || dbConfig.CacheConfig.RevCacheConfig.Size == nil {
+			dbConfig.CacheConfig.RevCacheConfig.Size = dbConfig.DeprecatedRevCacheSize
+		}
+		warnings = append(warnings, fmt.Sprintf(warningMsgFmt, "rev_cache_size", "cache.rev_cache.size"))
+	}
+
+	if dbConfig.CacheConfig != nil {
```
Comment (adamcfraser) [inline rest/config.go]: As discussed, don't need nil check here (and below)

**Review by adamcfraser — APPROVED** [review]

**Review by bbrks — APPROVED** [review]

**Pull request merged**